web-api/src/personalization/routes.rs

```diff
@@ -151,6 +167,154 @@ pub(crate) async fn update_interactions(
     Ok(())
 }
 
+#[derive(Deserialize)]
+struct StatelessPersonalizationRequest {
+    #[serde(default)]
+    count: Option<usize>,
+    #[serde(default)]
+    published_after: Option<DateTime<Utc>>,
+    history: Vec<UncheckedHistoryEntry>,
+}
+
+impl StatelessPersonalizationRequest {
+    fn resolved_history(
+        self,
+        config: &PersonalizationConfig,
+        warnings: &mut Vec<Warning>,
+    ) -> Result<Vec<HistoryEntry>, Error> {
+        let max_history_len = config.max_stateless_history_size;
+        if self.history.len() > max_history_len {
+            warnings.push(format!("history truncated, max length is {max_history_len}").into());
+        }
+        let mut most_recend_time = Utc::now();
+        //input is from oldest to newest
+        let mut history = self
+            .history
+            .into_iter()
+            .rev()
+            .take(max_history_len)
+            .map(|unchecked| -> Result<_, Error> {
+                let id = unchecked.id.try_into()?;
+                let timestamp = unchecked.timestamp.unwrap_or(most_recend_time);
+                if timestamp > most_recend_time {
+                    warnings
+                        .push(format!("inconsistent history ordering around document {id}").into());
+                }
+                most_recend_time = timestamp;
+                Ok(HistoryEntry { id, timestamp })
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        history.reverse();
+        Ok(history)
+    }
+
+    fn document_count(&self, config: &PersonalizationConfig) -> Result<usize, Error> {
+        let count = self.count.map_or(config.default_number_documents, |count| {
+            count.min(config.max_number_documents)
+        });
+
+        if count > 0 {
+            Ok(count)
+        } else {
+            Err(BadRequest::from("count has to be at least 1").into())
+        }
+    }
+}
+
+#[derive(Deserialize)]
+struct UncheckedHistoryEntry {
+    id: String,
+    timestamp: Option<DateTime<Utc>>,
+}
+
+#[derive(Deserialize)]
+struct HistoryEntry {
+    id: DocumentId,
+    timestamp: DateTime<Utc>,
+}
+
+#[instrument(skip_all)]
+async fn stateless_personalize_documents(
+    state: Data<AppState>,
+    Json(request): Json<StatelessPersonalizationRequest>,
+) -> Result<impl Responder, Error> {
+    let mut warnings = vec![];
+    let published_after = request.published_after;
+    let count = request.document_count(state.config.as_ref())?;
+    let history = request.resolved_history(state.config.as_ref(), &mut warnings)?;
+    let ids = history.iter().map(|document| &document.id).collect_vec();
+    let time = history
+        .last()
+        .map_or_else(Utc::now, |entry| entry.timestamp);
```

Comment:
:art: since we need to guarantee that `history` is not empty in `resolved_history()`, we can also unwrap here
```suggestion
    let time = history.last().copied().unwrap(/* history has been checked */);
```